src/extension.ts

```diff
@@ -0,0 +1,78 @@
+/**
+ * d2 File Viewer Extension
+ **/
+
+import {
+	commands,
+	ExtensionContext,
+	TextDocument,
+	TextDocumentChangeEvent,
+	window,
+	workspace,
+	WorkspaceConfiguration
+} from 'vscode';
+
+import { DocToPreviewGenerator } from './docToPreviewGenerator';
+
+const d2Ext = 'd2';
+
+const previewGenerator: DocToPreviewGenerator = new DocToPreviewGenerator();
+const ws: WorkspaceConfiguration = workspace.getConfiguration('d2-viewer');
+export let extContext: ExtensionContext;
+
+export function activate(context: ExtensionContext): void {
+
+	extContext = context;
+
+	context.subscriptions.push(workspace.onDidChangeTextDocument((e: TextDocumentChangeEvent) => {
+		if (e.document.languageId === d2Ext) {
+			const autoUp = ws.get('autoUpdate', false);
+
+			if (autoUp) {
+				const trk = previewGenerator.getTrackObject(e.document);
+				trk?.timer?.reset();
+			}
+		}
+	}));
+
+	context.subscriptions.push(workspace.onDidSaveTextDocument((doc: TextDocument) => {
+		if (doc.languageId === d2Ext) {
+			const updateOnSave = ws.get('updateOnSave', false);
+
+			if (updateOnSave) {
+				previewGenerator.generate(doc);
```

Comment:
Added change handler to avoid using cached values from the workspace.

Reply:
that's good but this one isn't about using cached values, it is to avoid the timer from rendering again if you have both settings on and save before the timer goes off.